### DISCUSSION TITLE
FastAPI tutorial returns "Unknown client"

### DISCUSSION THREAD

**Discussion by koaning** [fief-dev/fief#68] [Q&A]
I'm trying out [this tutorial](https://docs.fief.dev/integrate/python/fastapi/) but I seem to be stuck. When I click the **authorize** button from the FastAPI docs it leads me to: 

<img width="767" alt="image" src="https://user-images.githubusercontent.com/1019791/188327338-8a911aa7-4b55-4fc4-bcd5-6b3fbeac5147.png">

So that seems like it's broken? But I'm unsure if it's broken on my end or on your server. 

The tutorial might be better if it explained how to set the settings in the UI as well with some screenshots on where to get the settings. I think I've set it up correctly. 

<img width="309" alt="image" src="https://user-images.githubusercontent.com/1019791/188327465-db12f37c-6385-461a-bf8c-1eb0bd33383a.png">

```python
from fastapi import Depends, FastAPI
from fastapi.security import OAuth2AuthorizationCodeBearer
from fief_client import FiefAccessTokenInfo, FiefAsync
from fief_client.integrations.fastapi import FiefAuth

fief = FiefAsync(
    "base_url_from_tenant_menu",
    "45fu4...",
    "raMd...",
)
```

Did I misconfigure? Or was this a hiccup on your end?

**Comment by frankie567** [comment]
It looks like there is either something wrong with your base URL or your client ID. Several things to check:

* When you are redirected, the URL should look like this: `https://mytenant.fief.dev/authorize?response_type=code&client_id=CLIENT_ID&redirect_uri=http%3A%2F%2Flocalhost:8000%auth%2Fcallback&scope=openid`
* Clients are tied to one (and only one) tenant. So if you have several tenants, make sure the base URL corresponds to the client's tenant.

> The tutorial might be better if it explained how to set the settings in the UI as well with some screenshots on where to get the settings. I think I've set it up correctly.

There are more details here: https://docs.fief.dev/integrate/python/#create-a-fief-client

**Reply by frankie567** [reply]
Ok, I think I figured out what we missed here! You probably forgot to change the base URL on the scheme below:

```py
scheme = OAuth2AuthorizationCodeBearer(
    "https://example.fief.dev/authorize",  # Replace with your base URL here
    "https://example.fief.dev/api/token",  # And here
    scopes={"openid": "openid", "offline_access": "offline_access"},
)
```

Otherwise, the Swagger docs will redirect you to `example.fief.dev`, which obviously is not your tenant ;)

I improved the docs to heads up on this: https://docs.fief.dev/integrate/python/fastapi/#api-example

**Comment by koaning** [comment]
Good catch! I fixed the issue but hit my head against something else. I noticed that I was getting redirect errors. It seems that `http://localhost:8000/docs/oauth2-redirect` and `http://127.0.0.1:8000/docs/oauth2-redirect` are different URLs. After adding the `127.0.0.1` variant to the list as well I got a new issue though: when I log out and try to log back in I see this: 

<img width="600" alt="image" src="https://user-images.githubusercontent.com/1019791/188384775-0b7d741f-6f99-4b7b-96b0-0cbb1867a1a9.png">

Just to check, is this an issue on my side?

**Reply by frankie567** [reply]
Indeed `localhost` and `127.0.01` are considered different URL 😃 

As for the `invalid_grant` error, I think there is a bug in Swagger: it probably reuses the authorization code of the previous request. It doesn't happen though if you close/open the Authorize window again.

**Reply by koaning** [reply]
In order to prevent any issues related to the Swagger page: is there a demo that I can download that just uses Jinja2 with FastApi and an actual login form? That would make it easier to confirm how easy it might be integrate with my project.

**Reply by frankie567** [reply]
You should probably try the web application example: https://docs.fief.dev/integrate/python/fastapi/#web-application-example

**Reply by chrisorner** [reply]
I had the exact same issues as @koaning. In the tutorial it is described that the redirect_uri `http://localhost:8000/docs/oauth2-redirect` has to be enabled but actually the uri with `127.0.0` needs to be enabled.

**Reply by shirkan** [reply]
Hi,
So I came across the same issue, and after I added `http://127.0.0.1:8000/docs/oauth2-redirect` to redirect URIs it seems to be working, but only with the main authorization button and after checking at least one scope.
@frankie567 I think this is worth mentioning in the tutorial in [https://docs.fief.dev/integrate/python/fastapi/](https://docs.fief.dev/integrate/python/fastapi/) as others might fall here, just like me and others have.
Thanks!

**Reply by frankie567** [reply]
This is explained in the Getting Started tutorial: https://docs.fief.dev/getting-started/try-oauth2/
